rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,385 @@
+## Background
+
+So I'm working on improving the focus management and I came across an idea. However, not one that I felt comfortable just dropping a PR for without opening it up to discussion first (since it could just be a horrendous idea). So here's the problem I found and a possible solution I came up with.
+
+## Problem
+
+A lot of internal stuff is exposed to the end user that either doesn't need to be or shouldn't be. At best it can be confusing or misleading to users, and at worst it can cause some users to use code in ways they shouldn't (i.e., unintentional hacks that are unstable and could break between versions).
+
+For example, it seems wrong that a user can do this: `self.set_id(Index::default())`. And while this might not be that big of an issue ( "users should just *not* do that" ), what if we want to add other functionality onto widgets in the future? Maybe we add the ability to give widgets their own classes for theming. This might be confusing for the user who thinks that `self.add_class("foo")` should work on its own (but might not since the context also needs to  be made aware of the change).
+
+Additionally, `KayakContext`, has a lot of functionality that doesn't need to be exposed within the context of a widget, but needs to elsewhere (such as the `KayakContext::render` and `KayakContext::process_events` methods). Methods like these need to be public, but they're dangerous to use within a widget's render method.
+
+So in summary the two main issues are:
+
+1. Internal widget data is exposed to the user in the render function in a confusing manner
+2. Error-prone `KayakContext` methods are also exposed to the user in the render function when they shouldn't be
+
+## Solution
+
+I think the widget system could be restructured in a way that creates a layer of abstraction between the widgets and their context.
+
+Let's assume we have this widget:
+
+```rust
+#[widget]
+fn MyWidget(name: String, fill: Option<Color>) {
+  // ...
+}
+```
+
+### Props Extraction
+
+To get around the first issue of the widget having too much mutable access to its required widget data, we could separate the render function from the widget itself by placing the actual logic inside an associated method. 
+
+This would remove the access to a mutable `self`. But we still need access to the actual props. To solve this issue we can generate a new struct, `MyWidgetProps`, to contain the prop data. We can then pass this into the render function.
+
+And now we have something that looks a bit more like this:
+
+```rust
+struct MyWidget {
+  pub id: Index,
+  pub children: Children,
+  pub props: MyWidgetProps,
+  // ...
+}
+
+struct MyWidgetProps {
+  pub name: String,
+  pub fill: Option<Color>,
+}
+
+impl Widget for MyWidget {
+  fn render(&mut self, context: &mut KayakContext) {
+    Self::render(self.props, context);
+  }
+}
+
+impl MyWidget {
+  fn render(props: MyWidgetProps, context: &mut KayakContext) {
```

Comment:
I like this, but where do the props come from into the `render`?

Reply:
This should probably be called `render_internal` or something.

Also, this is only part of the proposed changes (I was trying to show my thought process leading up to the final proposal). In the end, this would actually look more like:

```rust
impl Widget for MyWidget {
  // ...
  fn render(&mut self, context: &mut KayakContext) {
    let mut widget_context = /* ... create MyWidgetContext ... */
    // Pass off to associated method so it loses access to dangerous MyWidget internals
    Self::render_internal(widget_context);
  }
}

impl MyWidget {
  fn render_internal(mut widget_context: MyWidgetContext) {
    // Render logic from functional component
  }
}
```

So it would be called by the `MyWidget` instance, specifically in its implementation of the `Widget::render` trait method:

`KayakContext` -(via `Box<dyn Widget>`)-> `MyWidget::render(&mut self, ...)` -> `MyWidget::render_internal()`

Reply:
I think this makes sense. How would this work for people who want to not use `rsx!` though? 🤔

Reply:
My thought was that this is all for users who want to write functional components (i.e. they don't want to write their own super-custom logic or boilerplate). And for users who prefer to write it all out manually, then they'll have to write their logic on their own (which isn't a bad thing if the point is they want more control over functionality).

And since `render_internal`, `MyWidgetContext`, etc. are all generated by the macro, they don't need to implement or recreate any of it. They can do pretty much what we have now for [`Fragment`](https://github.com/StarArawn/kayak_ui/blob/dfdee36240be1236081c480cb8f573ab6c43e586/kayak_core/src/fragment.rs) and [`VecTracker`](https://github.com/StarArawn/kayak_ui/blob/dfdee36240be1236081c480cb8f573ab6c43e586/kayak_core/src/vec.rs).

So if they want to "add a class for theming", they'd have to do something like:

```rust
impl MyCustomWidget {
  fn add_class(&mut self, context: &mut KayakContext, class_name: &str) {
    println!("Added the class: {}", class_name);
    self.class_list.insert(class_name.to_string());
    context.add_class(self.id, class_name);
    // Other custom stuff...
  }
}

impl Widget for MyCustomWidget {
  fn render(&mut self, context: &mut KayakContext) {
    // Do custom render logic (as we do now basically)
    self.add_class(context, "foo");
    // ...
  }
}
```